osu.Game/Screens/Select/BeatmapCarousel.cs

```diff
@@ -154,6 +162,37 @@ private void load(OsuConfigManager config)
             beatmaps.BeatmapRestored += beatmapRestored;
 
             loadBeatmapSets(GetLoadableBeatmaps());
+
+            decoupledRuleset.BindValueChanged(UpdateRecommendedStarDifficulty, true);
+        }
+
+        protected void UpdateRecommendedStarDifficulty(ValueChangedEvent<RulesetInfo> ruleset)
+        {
+            if (api.LocalUser.Value is GuestUser)
+            {
+                RecommendedStarDifficulty.Value = 0;
+                return;
+            }
+
+            var req = new GetUserRequest(api.LocalUser.Value.Id, ruleset.NewValue);
+
+            req.Success += result =>
+            {
+                // algorithm taken from https://github.com/ppy/osu-web/blob/e6e2825516449e3d0f3f5e1852c6bdd3428c3437/app/Models/User.php#L1505
+                RecommendedStarDifficulty.Value = Math.Pow((double)(result.Statistics.PP ?? 0), 0.4) * 0.195;
+            };
+
+            api.PerformAsync(req);
+        }
+
+        internal bool IsRecommended(BeatmapInfo beatmap)
+        {
+            if (RecommendedStarDifficulty.Value == 0)
+                return false;
+
+            BeatmapSetInfo set = beatmap.BeatmapSet;
+            BeatmapInfo recommended = set.Beatmaps.OrderBy(b => SongSelect.DifferenceToRecommended(b, RecommendedStarDifficulty.Value)).FirstOrDefault();
+            return beatmap == recommended && Math.Abs(beatmap.StarDifficulty - RecommendedStarDifficulty.Value) < 1.5;
```

Comment:
This should actually be ``< 0.5``

Found in [osu-web](https://github.com/ppy/osu-web/blob/4687a1e8a5c132af3ccfc63a2b90d239ec395d71/app/Libraries/Search/BeatmapsetSearch.php#L174-L176)